pkg/codegen/hcl2/model/type_enum.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// EnumType represents values of a single type, and a closed set of possible values.
+type EnumType struct {
+	// Elements are the possible values of the enum.
+	Elements []cty.Value
+
+	// The type of the enum's values.
+	Type Type
+
+	// The associated schema type. It is necessary to include this, since
+	// otherwise identical types are distinct if they have a different schema
+	// type.
+	Token string
```

Comment:
The types in `hcl2/model` aren't intended to deal with schema types directly. Instead, the intent is that schema types are associated with `hcl2/model` types via annotations (for example, see the `Annotations` field in [`model.ObjectType`](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/hcl2/model/type_object.go#L30-L44) and the way it is used [by the `pcl` package](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/pcl/binder_schema.go#L302-L305)). Although the annotations approach is a little janky, I'd prefer that we remain consistent throughout this package until we find a better alternative.

Reply:
I'm good with moving the `src *schema.EnumType` field into an annotation. It's a direct reference to the schema (at the type level), and while it's necessary for codegen, it is not inherently necessary for specifying a `EnumType`. That makes sense to me. 

The `Token` (I'm fine calling it something else, GUID comes to mind) is an intrinsic part of defining an `EnumType`, because `EnumType {Elements: [1,2], Type: Number, Token: "foo"}` is an entirely different type then `EnumType {Elements: [1,2], Type: Number, Token: "bar"}`. It seems very strange to have basic methods like `Equals` rely on an untyped annotation. I can ensure that we don't rely on any property of `Token` in `hcl2/model` besides uniqueness.

Reply:
> The Token (I'm fine calling it something else, GUID comes to mind) is an intrinsic part of defining an EnumType, because EnumType {Elements: [1,2], Type: Number, Token: "foo"} is an entirely different type then EnumType {Elements: [1,2], Type: Number, Token: "bar"}

So I'm not sure that I agree with the premise--there's nothing about enum types in the abstract that requires them to have nominative semantics--but I see how it's helpful for PCL. Let's ship this as-is and consider following up with separate nominative types later (which might also be useful as a replacement for the fragile annotations system).

Reply:
That makes sense. I think I need to read more about type systems. I tend to think the value of an `Enum` as its name, but that isn't the only way to think about it. I'm curious what you imagine. Something like a `Named<T>` wrapper type? 

I'll leave a comment suggesting the refactor.